### ISSUE TITLE
Access violation (C0000005) detected

### ISSUE THREAD

**Issue by ivarbergman** [Starcounter/Starcounter#3699]
Hi! On one of our production server we get an access violation error on Launcher start. 
We are using SC Version=2.2.1.2230

Can you help us understand if there are some problems with our database, or if there is something else we need to do to fix this problem?

(Since this is on our production server, the priority on this issue is high.)

```
Access violation (C0000005) detected at ..\..\src\sccoredb\sccoredb\sccoredb.cpp:424 (Thread 5636).
    00000090B6F71663 sccoredb!get_slot_read_ptr : getslotreadptr.hpp (35)
    00000090B6F6B48E sccoredb!__get_table_info : schemaapi.cpp (660)
    00000090B6F5EB99 sccoredb!sccoredb_get_table_info : sccoredb.cpp (415)
    000007FF688916CB <unknown>!
    000007FF688B0142 <unknown>!
    00000090F64CEFF0 <unknown>!
    00000090F64CF030 <unknown>!
    0000008FD8BC22C0 <unknown>!
    00008FED48055A0E <unknown>!
    000007FFC85654B0 clr!
    FFFFFFFFFFFFFFFF <unknown>!
    Endless stackwalk detected.
    An error occured during callstack dump.
```

**Comment by ivarbergman** [comment]
You can download a zip archive of the Data files here:

http://tunitybuild.cloudapp.net:9999/issue-3699-intranet.zip

**Comment by per-samuelsson** [comment]
Hi @ivarbergman,

Sorry to see you experience this. It looks like it's something already in our pipe, considering older issues with similar stack, such as #3545.

Let's see if we can first establish that. @eriohl?

**Comment by eriohl** [comment]
I'll check the database and see if I can see what is wrong.

**Comment by eriohl** [comment]
At first glance it looks like the same problem reported before. I will try recreating the database to see if I can reproduce when the problem occurs.

**Comment by eriohl** [comment]
It seems that there is still some with table updates unfortunately. I can reproduce it but it will probably take some time to address.

**Comment by ivarbergman** [comment]
Hi! Thanks for the feedback. A couple of questions: 

Is there something in our code that causes these errors? (Ie something we need to fix.) I have seen a number of 

    ScErrStackOverflow (SCERR1011): A stack overflow has been detected.

errors in the starcounter log for this database. Could circular references between database objects be the  reason for these Access violation errors? 

When you investigated our database files, was it some specific database class that caused the error? 

Also, I need to inform the Tunity team on when we can have this server up and running again. Do you have a rough time estimate on when a fix of the "table updates" problems can be expected?

Or, If you have some suggestion what we can do to avoid the start up error. That would be much appreciated.

Thanks!

**Comment by eriohl** [comment]
I found the problem. Embarrassingly enough it is another bug that causes pretty much the same problem as in the case #3532.

Unfortunately the bug has messed the image file. But since I have all the log I can rebuild it for you. Then you should be able to just swap out the database and start as normal. Would that work for you?

**Comment by ivarbergman** [comment]
Super!! 

Would it be possible for you to rebuild the images before tomorrow morning? That would be awesome :)

Will we need to upgrade SC to some specific version or will your "cleaned" image files be sufficient?

**Comment by eriohl** [comment]
Rebuilding now. Should not take long. A few minutes if I got the configuration right this time.

4 GiB RAM on your server right?

Do you have anywhere I can put the database when done?

> Will we need to upgrade SC to some specific version or will your "cleaned" image files be sufficient?

You need a new version for the underlying bug to be fixed. But you should be able to start with the new image file with the current version.

**Comment by eriohl** [comment]
Got a rebuilt database for you.

**Comment by eriohl** [comment]
Fix pushed. Will be included in next build (2.2.1.2535).

I'll make sure a new version is available as soon as possible.

**Comment by ivarbergman** [comment]
Great! We are running Azure A2 instances (3.5 GB RAM) it has worked fine so far. Do you think that we need to increase RAM?

**Comment by eriohl** [comment]
> Great! We are running Azure A2 instances (3.5 GB RAM) it has worked fine so far. Do you think that we need to increase RAM?

No. Just asking to get the database configuration correct.

**Comment by ivarbergman** [comment]
Startup now gets past the Access violation problem, but it looks like I have another problem with my database.

```
System.Exception: ScErrUnexpMetadata (SCERR7035): Internal error in unmanaged meta-data population. Unexpecably not found metadata for column Type in table Concepts.Ring8.Tunity.OrderNumber
Version: 2.2.1.2230.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR7035.
   at Starcounter.SqlProcessor.MetadataPopulation.<>c__DisplayClass0_0.<PopulateClrMetadata>b__0() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-324\Level1\src\Starcounter\SqlProcessor\MetadataPopulation.cs:line 16
   at Starcounter.Db.Transact(Action action, UInt32 flags, Boolean forceSnapshot, Int32 maxRetries) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-324\Level1\src\Starcounter\Db.cs:line 355
   at Starcounter.Hosting.Package.UpdateDatabaseSchemaAndRegisterTypes(String fullAppId, TypeDef[] unregisteredTypeDefs, TypeDef[] allTypeDefs) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-324\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 353
   at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application, ApplicationDirectory applicationDir) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-324\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 274
   at Starcounter.Hosting.Package.Process() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-324\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 138
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-324\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 71
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR7035

```

It should be mentioned that the class `Concepts.Ring8.Tunity.OrderNumber` is added in the latest update of our apps.

(I will send you our  binaries shortly.)

/Ivar

**Comment by ivarbergman** [comment]
Here is a zip with our binaries. Most likely it should have been enough to test with `Launcher.exe` since start up fails on this app.

http://tunitybuild.cloudapp.net:9999/issue-3699_app_binaries.zip

Hope this helps!

**Comment by eriohl** [comment]
cc @k-rus @alexanderstc Any insight?

**Comment by alexanderstc** [comment]
This is develop, right - @k-rus is the man.

**Comment by eriohl** [comment]
I can reproduce (with a version including the fix). Just running Launcher.exe with the rebuilt database.

**Comment by k-rus** [comment]
If I don't mistake the exception happens in MetadataPopulation.cs line 76. I think as work around the population of CLR metadata can be disabled without breaking functionality of this version of Starcounter. Some our regression test will fail. If to run daily develop, then it should succeed I think. @eriohl can you try, since you have environment in place? Just comment line calling `PopulateClrMetadata`.

**Comment by eriohl** [comment]
Sure I can try that. But I don't see why that would really help.

There problem seems to be that there is a Starcounter.Metadata.Table record for OrderNumber but no entries in the Starcounter.Metadata.Column table for the table (ie the metadata for the class is incomplete). Can't I just reset it for example removing the record in Starcounter.Metadata.Table so that the metadata is recreated properly or something?

**Comment by k-rus** [comment]
> Can't I just reset it for example removing the record in Starcounter.Metadata.Table so that the metadata is recreated properly or something?

I don't think it will help, since if stored table definition is equal to the definition coming from class, then no meta-data population code for the class is called.

**Comment by eriohl** [comment]
I managed to get it to create the Starcounter.Metadata.Column records for OrderNumber by fooling it that the table has been updated. But then I got the same problem with a table called FormData. After doing the same with that table it starts.

Not sure on how to verify that everything works properly after that however....

**Comment by k-rus** [comment]
If I remember correctly, in develop metadata is not used anywhere.

**Comment by ivarbergman** [comment]
The FormData class was added in the same app update as OrderNumber. I forgot about that class. 

@eriohl  Can you upload the Data images you have used when you successfully started the Launcher? I can give it another try on our server.

Thanks!

**Comment by eriohl** [comment]
As far as I can tell my tampering created the metadata table entries for the columns and nothing else. So it shouldn't have caused any problems. But I will double check that it didn't do anything strange.

@ivarbergman : I've uploaded the database to your FTP: issue-3699-intranet-rebuild2.zip

There is also a new nightly with the fix included: 2.2.1.2540. I would recommend you upgrade as well although I think it will run regardless since the upgrade should not be completed.

**Comment by k-rus** [comment]
@eriohl How do I build Level1, develop nowadays? I remember I've seen something recently that requires some manual work. I have problem:

>NuGet Package restore failed for project 3a. MsBuild\Starcounter.XSON.PartialClassGenerator

**Comment by k-rus** [comment]
> As far as I can tell my tampering created the metadata table entries for the columns and nothing else. So it shouldn't have caused any problems. 

Great! This approach is better than disabling metadata, since it does not require to use specialised build of Starcounter with disabled metadata population. (I tested that disabling metadata works, see [this build](https://teamcity.starcounter.org/viewLog.html?buildId=5316&tab=buildResultsDiv&buildTypeId=Starcounter_DevelopDaily)

> But I will double check that it didn't do anything strange.

For the safety, after starting production database the data can be unloaded and then loaded into newly created database. As usual it is necessary that all apps are loaded. See [detailed instructions](https://github.com/Starcounter/Starcounter/wiki/Reloading-database-between-Starcounter-versions)

**Comment by eriohl** [comment]
Yes. I had this problem as well. NuGet config was wrong so it couldn't download the packages.

In the upper right corner of "Manage NuGet packages for this solution" there is a package source which was empty. Click on the cogwheel and check the one that says "Microsoft and .NET". After that package restore worked.

**Comment by k-rus** [comment]
> Yes. I had this problem as well. NuGet config was wrong so it couldn't download the packages.
> 
> In the upper right corner of "Manage NuGet packages for this solution" there is a package source which was empty. Click on the cogwheel and check the one that says "Microsoft and .NET". After that package restore worked.

Thank you! The instruction worked great! Now I am able to build Level1 again.

**Comment by ivarbergman** [comment]
@eriohl our server is back online!! Latest images work perfekt! Thanks!!

**Comment by dan31** [comment]
Thank you @eriohl @k-rus @ivarbergman @per-samuelsson for operativeness! @k-rus, should we assume your commit alongside this issue is a fix (to get included in upcoming stable release)?

**Comment by eriohl** [comment]
> As far as I can tell my tampering created the metadata table entries for the columns and nothing else. So it shouldn't have caused any problems. But I will double check that it didn't do anything strange.

I've double checked and the only additional transactions applied are metadata updates. Nothing looks odd from what I can see. Doesn't mean that metadata is all as it should now however. This I can't tell. So maybe taking @k-rus advice and rebuilding the database is prudent.

> @eriohl our server is back online!! Latest images work perfekt! Thanks!!

Great! Thanks for all your help and patience.

**Comment by eriohl** [comment]
> ..  @k-rus, should we assume your commit alongside this issue is a fix (to get included in upcoming stable release)?

Underlying problem seems to be that the Starcounter.Metadata.Column entries for the new tables was not created properly (they are created in a separate transaction). I think because the initial error took down the database before they could be created and after that, since the tables was already created, it saw no need. Disabling creation of these CLR metadata tables does not solve the underlying problem. That being said I'm not sure what effect the lack of these records actually would have other then the reported process failure.

Anyway: I've put the broken and fixed database on company dropbox: Exchange\Erik.Ohlsson: issue-3699-intranet-rebuild.zip (fixed image) issue-3699-intranet-rebuild2.zip (fixed metadata) if you want to take a look @k-rus (I hope this is ok @ivarbergman? Only for internal use debugging. Let us know is not and we'll delete them promptly).

I think to fix the problem one needs to make sure the Starcounter.Metadata.Column entries are there regardless of if the table has been upgraded or not. Since they are created in a different transaction there is always a risk that they are not properly created.

**Comment by eriohl** [comment]
> Is there something in our code that causes these errors? (Ie something we need to fix.) I have seen a number of

> ```ScErrStackOverflow (SCERR1011): A stack overflow has been detected.```

> errors in the starcounter log for this database. Could circular references between database objects be the reason for these Access violation errors?

> When you investigated our database files, was it some specific database class that caused the error?

On this I can't advice unfortunately other then to say that it is unrelated to this problem. This message only says that a stack overflow has occurred in the host. It provides no additional information nor affect the result of the exception.

The reason why this output was added was because the CLR had a policy of silently dying when this occurred unless there was a debugger attached. To actually find the problem we had to attach a debugger to the production server process so that the error could be intercepted. But this was CLR 2.0  and a lot has changed since then. I personally have not worked a lot with the CLR since then however.

Maybe @per-samuelsson or @chrhol can advice on this?

**Comment by per-samuelsson** [comment]
> NuGet Package restore failed for project 3a. MsBuild\Starcounter.XSON.PartialClassGenerator

I've never seen that one reported. It's likely something related to us now using Roslyn. I'm opening up an admin track issue about it to see if we can improve.

**Comment by k-rus** [comment]
> @k-rus, should we assume your commit alongside this issue is a fix (to get included in upcoming stable release)?

I didn't fix anything. I just tested a work around to start the database.



> Underlying problem seems to be that the Starcounter.Metadata.Column entries for the new tables was not created properly (they are created in a separate transaction). I think because the initial error took down the database before they could be created and after that, since the tables was already created, it saw no need. Disabling creation of these CLR metadata tables does not solve the underlying problem. That being said I'm not sure what effect the lack of these records actually would have other then the reported process failure.

Thank you for the investigation. I will look at it for a possible fix. Metadata are not crucial in branch develop, however they play more important role in branch pnext.

**Comment by per-samuelsson** [comment]
> Maybe @per-samuelsson or @chrhol can advice on this?

If that refers to the occurrence of the stack overflow, I know for a fact we have a know case where it happens in Typed JSON, that @chrhol is aware of, but where there is no clear way to fix it: https://github.com/Starcounter/Starcounter/issues/3147. That's my best guess, but hard to say for sure when we don't have anything else to go on.

Maybe @ivarbergman can give some insight when things have stabilized a bit.

**Comment by eriohl** [comment]
Closing this one since no additional input.

**Comment by k-rus** [comment]
> Underlying problem seems to be that the Starcounter.Metadata.Column entries for the new tables was not created properly (they are created in a separate transaction). I think because the initial error took down the database before they could be created and after that, since the tables was already created, it saw no need. Disabling creation of these CLR metadata tables does not solve the underlying problem. That being said I'm not sure what effect the lack of these records actually would have other then the reported process failure.
> Thank you for the investigation. I will look at it for a possible fix. Metadata are not crucial in branch develop, however they play more important role in branch pnext.

Putting population of properties into the same transaction seems to be not possible due to limitation that tables should be created in separate internal transaction.

However, it is possible to check and repair meta-data, but it is probably not worth it, since version 2.3 has different implementation for DDLs and meta-data population (and own issue on meta-data repair). So I suggest to skip fix for 2.2 versions. Let me know if it is actual critical to fix.
